File: bin/jobs.rs
use std::path::{Path, PathBuf};

use super::harness::{SOURCE_AGENT_FILE, SyncEnv};
use super::managed::asset_dir_names;
use super::{copy_tree, err, is_symlink, panic_message, rm_entry};

#[derive(Clone, Copy, Debug)]
pub(super) enum JobKind {
    File,
    Dir,
}

#[derive(Clone, Debug)]
pub(super) struct Job {
    src: PathBuf,
    dst: PathBuf,
    kind: JobKind,
}

pub(super) fn copy_item(src: &Path, dst: &Path) -> bool {
    if !src.exists() && !is_symlink(src) {
        err(&format!("missing source: {}", src.display()));
        return true;
    }

    if let Some(parent) = dst.parent() {
        std::fs::create_dir_all(parent).unwrap_or_else(|error| panic!("{error}"));
    }
    rm_entry(dst).unwrap_or_else(|error| panic!("{error}"));

    let result = if src.is_dir() {
        copy_tree(src, dst)
    } else {
        std::fs::copy(src, dst).map(|_| ())
    };

    if let Err(error) = result {
        err(&format!(
            "copy failed: {} -> {} ({error})",
            src.display(),
            dst.display()
        ));
        return false;
    }

    true
}

pub(super) fn copy_dir_into(src_dir: &Path, dst_dir: &Path) -> bool {
    if !src_dir.is_dir() {
        err(&format!("missing directory: {}", src_dir.display()));
        return true;
    }

    std::fs::create_dir_all(dst_dir).unwrap_or_else(|error| panic!("{error}"));
    if let Err(error) = copy_tree(src_dir, dst_dir) {
        err(&format!(
            "copy failed: {} -> {} ({error})",
            src_dir.display(),
            dst_dir.display()
        ));
        return false;
    }

    true
}

pub(super) fn iter_jobs(sync_env: &SyncEnv) -> Result<Vec<Job>, String> {
    let mut jobs = harness_dirs(sync_env);
    jobs.extend(asset_copies(sync_env)?);
    jobs.extend(agent_files(sync_env));
    jobs.extend(config_files(sync_env));
    Ok(jobs)
}

pub(super) fn run_jobs(jobs: &[Job]) -> bool {
    jobs.iter().all(run_job)
}

fn run_job(job: &Job) -> bool {
    let (name, handler): (&str, fn(&Path, &Path) -> bool) = match job.kind {
        JobKind::Dir => ("copy_dir_into", copy_dir_into),
        JobKind::File => ("copy_item", copy_item),
    };

    match std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| handler(&job.src, &job.dst))) {
        Ok(result) => result,
        Err(payload) => {
            err(&format!(
                "unexpected error in {name}: {}",
                panic_message(payload)
            ));
            false
        }
    }
}

fn harness_dirs(sync_env: &SyncEnv) -> Vec<Job> {
    sync_env
        .harnesses
        .iter()
        .map(|harness| Job {
            src: harness.source_root(&sync_env.tools_home),
            dst: harness.root(),
            kind: JobKind::Dir,
        })
        .collect()
}

fn asset_copies(sync_env: &SyncEnv) -> Result<Vec<Job>, String> {
    let asset_names = asset_dir_names(&sync_env.assets_home)?;
    let mut jobs = Vec::new();
    for asset_name in asset_names {
        let asset_path = sync_env.assets_home.join(&asset_name);
        for harness in &sync_env.harnesses {
            jobs.push(Job {
                src: asset_path.clone(),
                dst: harness.root().join(harness.rename_asset(&asset_name)),
                kind: JobKind::Dir,
            });
        }
    }
    Ok(jobs)
}

fn agent_files(sync_env: &SyncEnv) -> Vec<Job> {
    sync_env
        .harnesses
        .iter()
        .map(|harness| Job {
            src: sync_env.assets_home.join(SOURCE_AGENT_FILE),
            dst: harness.instruction_target(),
            kind: JobKind::File,
        })
        .collect()
}

fn config_files(sync_env: &SyncEnv) -> Vec<Job> {
    vec![Job {
        src: sync_env.assets_home.join("mcporter.jsonc"),
        dst: sync_env.mcporter_home.join("mcporter.json"),
        kind: JobKind::File,
    }]
}

File: bin/jobs.ts
import fs from "node:fs";
import { dirname, join } from "node:path";

import { SOURCE_AGENT_FILE, SyncEnv } from "./harness.ts";
import { assetDirNames } from "./managed.ts";
import { copyTree, err, isSymlink, panicMessage, rmEntry } from "./lib.ts";

export type JobKind = "File" | "Dir";

export interface Job {
  src: string;
  dst: string;
  kind: JobKind;
}

export function copyItem(src: string, dst: string): boolean {
  try {
    if (!fs.existsSync(src) && !isSymlink(src)) {
      err(`missing source: ${src}`);
      return true;
    }

    fs.mkdirSync(dirname(dst), { recursive: true });
    rmEntry(dst);

    if (isDirectoryLike(src)) {
      copyTree(src, dst);
    } else {
      fs.copyFileSync(src, dst);
    }

    return true;
  } catch (error) {
    err(`copy failed: ${src} -> ${dst} (${panicMessage(error)})`);
    return false;
  }
}

export function copyDirInto(srcDir: string, dstDir: string): boolean {
  try {
    if (!isDirectoryLike(srcDir)) {
      err(`missing directory: ${srcDir}`);
      return true;
    }

    fs.mkdirSync(dstDir, { recursive: true });
    copyTree(srcDir, dstDir);
    return true;
  } catch (error) {
    err(`copy failed: ${srcDir} -> ${dstDir} (${panicMessage(error)})`);
    return false;
  }
}

export function iterJobs(syncEnv: SyncEnv): Job[] {
  const jobs: Job[] = [];
  jobs.push(...harnessDirs(syncEnv));
  jobs.push(...assetCopies(syncEnv));
  jobs.push(...agentFiles(syncEnv));
  jobs.push(...configFiles(syncEnv));
  return jobs;
}

export function runJobs(jobs: readonly Job[]): boolean {
  return jobs.every((job) => runJob(job));
}

function runJob(job: Job): boolean {
  try {
    return job.kind === "Dir"
      ? copyDirInto(job.src, job.dst)
      : copyItem(job.src, job.dst);
  } catch (error) {
    err(`unexpected error in ${job.kind === "Dir" ? "copy_dir_into" : "copy_item"}: ${panicMessage(error)}`);
    return false;
  }
}

function harnessDirs(syncEnv: SyncEnv): Job[] {
  return syncEnv.harnesses.map((harness) => ({
    src: harness.sourceRoot(syncEnv.toolsHome),
    dst: harness.root(),
    kind: "Dir",
  }));
}

function assetCopies(syncEnv: SyncEnv): Job[] {
  const jobs: Job[] = [];
  for (const assetName of assetDirNames(syncEnv.assetsHome)) {
    const assetPath = join(syncEnv.assetsHome, assetName);
    for (const harness of syncEnv.harnesses) {
      jobs.push({
        src: assetPath,
        dst: join(harness.root(), harness.renameAsset(assetName)),
        kind: "Dir",
      });
    }
  }
  return jobs;
}

function agentFiles(syncEnv: SyncEnv): Job[] {
  return syncEnv.harnesses.map((harness) => ({
    src: join(syncEnv.assetsHome, SOURCE_AGENT_FILE),
    dst: harness.instructionTarget(),
    kind: "File",
  }));
}

function configFiles(syncEnv: SyncEnv): Job[] {
  return [
    {
      src: join(syncEnv.assetsHome, "mcporter.jsonc"),
      dst: join(syncEnv.mcporterHome, "mcporter.json"),
      kind: "File",
    },
  ];
}

function isDirectoryLike(path: string): boolean {
  try {
    return fs.statSync(path).isDirectory();
  } catch {
    return false;
  }
}
